src/articles/2024-01-21_DDEV-for-js.md

```diff
@@ -0,0 +1,327 @@
+Starting Mutagen sync process...
+Mutagen sync flush completed in 1s.
+For details on sync status 'ddev mutagen st my-project -l'
+Waiting for web/db containers to become ready: [web db]
+^[Starting ddev-router if necessary...
+ Container ddev-router  Running
+Waiting for additional project containers to become ready...
+All project containers are now ready.
+Successfully started my-project
+Project can be reached at https://my-project.ddev.site https://127.0.0.1:33125
+```
+
+## Configuring for non-PHP projects
+
+While our project is running, there's some configuration we need to change. Open `my-project/.ddev/config.yml`:
+
+```yml
+name: my-project
+type: php
+docroot: ""
+php_version: "8.1"
+webserver_type: nginx-fpm
+xdebug_enabled: false
+additional_hostnames: []
+additional_fqdns: []
+database:
+  type: mariadb
+  version: "10.4"
+use_dns_when_possible: true
+composer_version: "2"
+web_environment: []
+```
+
+### Add additional hostnames
+
+DDEV is _highly_ configurable. All the key/value options are [documented here](https://ddev.readthedocs.io/en/latest/users/configuration/config). There are two main things we want to edit here. We're first going to add additional hostnames. Since keystone will be running our backend, we'll have it accessible at `api.my-project.ddev.site`, and since Sveltekit is running our frontend app, we'll have itat `app.my-project.ddev.site`. The default URL will also exist at `my-project.ddev.site`, but DDEV's container healthcheck needs to be able to access php on that URL, so we'll leave it alone.
+
+```yml
+additional_hostnames:
+  - "api.my-project"
+  - "app.my-project"
+```
+
+### Standardize your node version
+
+Next, we want to specify our Node version. This can be either a generic major version like `20` or a specific release like `16.4.2`. Since we want to avoid the "it works on my machine" problems, this value should match what you're running in production.
+
+```yml
+nodejs_version: "20"
+```
+
+### Customize your webserver config
+
+The next thing we need to do is override how DDEV has configured the webserver. Right now, _all_ requests coming into the `web` container are sent to `/my-project/` to look for an `index.html` or `index.php` file, regardless of URL. We'll need to add our nginx or apache configuration files (depending on your config.yml file) to proxy requests to the correct ports in the web container. Your default nginx configuration is in `.ddev/nginx_full/nginx-site.conf` and your default apache configuration is in `.ddev/apache/apache-site.conf`. Both of these files are auto-generated and we'll leave them alone. In both cases, there's a file `seconddocroot.conf.example` that we'll edit instead. Since we have two additional URLs (`app.my-project.ddev.site` and `api.my-project.ddev.site`), we'll need to make a config file for each. In the examples below, I've only included the files for the `app` subdomain.
+
+#### Configuring Nginx
```

Comment:
Make clear that you *either* do nginx *or* apache, don't have to do both.

Reply:
I think it's worth explaining a little bit about what we're doing here with proxy_pass and friends... sending certain types of requests to the Node.js server.